File: backend/pkg/db/session.go
package db

// Session holds all data about a session
type Session struct {
	ID       *int64    `json:"id" example:"1"`
	Timeslot *Timeslot `json:"timeslot"`
	Name     *string   `json:"name" example:"Session Name"`
	Speaker  *Speaker  `json:"speaker"`
	Room     *Room     `json:"room"`
}

// NewSession makes a new Session with default values
func NewSession() Session {
	room := NewRoom()
	speaker := NewSpeaker()
	timeslot := NewTimeslot()
	return Session{
		ID:       Int64Ptr(0),
		Timeslot: &timeslot,
		Name:     StringPtr(""),
		Speaker:  &speaker,
		Room:     &room,
	}
}

// SessionReaderWriterUpdaterDeleter implements everything that a facade for a Session would need
type SessionReaderWriterUpdaterDeleter interface {
	SessionReader
	SessionWriter
	SessionUpdater
	SessionDeleter
}

// SessionReader implements all read related methods
type SessionReader interface {
	ReadASession(sessionID int64) (Session, error)
	ReadAllSessions() ([]Session, error)
}

// SessionWriter implements all write related methods
type SessionWriter interface {
	WriteASession(speakerID *int64, roomID *int64, timeslotID *int64, name *string) (int64, error)
}

// SessionUpdater implements all update related methods
type SessionUpdater interface {
	UpdateASession(sessionID int64, speakerID *int64, roomID *int64, timeslotID *int64, name *string) error
}

// SessionDeleter implements all delete related methods
type SessionDeleter interface {
	DeleteASession(sessionID int64) error
}

File: frontend/src/app/data_models/session.ts
import { Room } from "./room";
import { Speaker } from "./speaker";
import { Timeslot } from "./timeslot";

export class Session {
  id: number;
  name: string;
  room: Room;
  speaker: Speaker;
  timeslot: Timeslot;

  constructor(name: string, room: Room, speaker: Speaker, timeslot: Timeslot) {
    this.id = -1;
    this.name = name;
    this.room = room;
    this.speaker = speaker;
    this.timeslot = timeslot;
  }
}
